File: Algorithm-CPP/45_공주구하기.cpp
#include <iostream>
using namespace std;

int main() {
    int n, k;
    cin >> n >> k;

    int arr[n];
    for(int i = 0; i < n; i++) arr[i] = 0;

    int index = 0;
    for(int i = 0; i < n - 1; i++) {
        int cnt = 1;

        while(1) {
            if(cnt == k && arr[index] == 0) {
                arr[index] = 1;
                break;
            } else if(cnt == k && arr[index] == 1) {
                index++;
            } else {
                if(arr[index] == 0) {
                    index++; cnt++;
                } else if(arr[index] == 1) index++;
            }

            if(index == n) index = 0;
        }

        arr[index] = 1;
    }

    for(int i = 0; i < n; i++) {
        if(arr[i] == 0) {
            cout << i + 1 << endl;
            return 0;
        }
    }
}

File: Algorithm-Swift/Algorithm-Swift/Inflearn/45_공주구하기.swift
//
//  45_공주구하기.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/22.
//

import Foundation

func solution45() {
    let nk = readLine()!.split(separator: " ").map { Int($0)! }
    let n = nk.first!, k = nk.last!
    var arr = Array(repeating: 0, count: n)
    var index = 0

    for _ in 0..<n-1 {
        var cnt = 1

        while true {
            if arr[index] == 0 && cnt == k {
                arr[index] = 1
                break
            } else if arr[index] == 1 && cnt == k {
                index += 1
            } else {
                if arr[index] == 0 {
                    index += 1
                    cnt += 1
                } else {
                    index += 1
                }
            }

            if index == n { index = 0 }
        }
    }

    arr.enumerated().forEach { index, num in
        if num == 0 {
            print(index + 1)
            return
        }
    }
}
